spring-kafka/src/main/java/org/springframework/kafka/listener/TimestampedException.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2018-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener;
+
+import java.time.Clock;
+import java.time.Instant;
+
+import org.springframework.core.NestedRuntimeException;
+
+/**
+ * A {@link NestedRuntimeException} that records the timestamp
+ * of when it was thrown.
+ *
+ * @author Tomaz Fernandes
+ * @since 2.7
+ */
+public class TimestampedException extends NestedRuntimeException {
```

Comment:
Should extend `KafkaException` (which is a `NestedRuntimeException`).

Reply:
> I concur with @artembilan - it is very difficult to review such large PRs, especially so close to a release. Given that this feature is still "experimental", I am ok with merging this with a less than thorough review, especially if we release RC2 in a couple of weeks.
> 
> Hopefully, this is the last large refactoring and we can look forward to more compact PRs in future.
> 
> But, once again @tomazfernandes , thanks for the great work.

Sounds good to me @garyrussell, I've made the adjustments you asked.

I'll sure break it down in smaller PRs in the future.

Let me know if there's anything else to be changed - for this PR or for RC2 if it's to be released.

Thanks!